Returns 404 if Grafana is only available on a specific URL path

I'm currently using a Grafana instance that's available under the url `host://grafana/`
This results in grafanadata returning `grafana returned status 404`

It looks like the query generation in `GrafanaClient.getDashboard(uid)`  doesn't account for this scenario yet:
```
	host := fmt.Sprintf("%v://%v", c.baseURL.Scheme, c.baseURL.Host)
	query := fmt.Sprintf("%v/api/dashboards/uid/%v", host, uid)
```

This url setup also breaks the extraction of the panel uid in the `ExtractArgs` function

Thank you for your response.
I did the following steps to validate the test:
1. Create a new Grafana instance running `docker run -d --name=grafana -p 80:3000 grafana/grafana`
2. Create Service Account with access token 
2. Create a test datasource
3. Create a simple dashboard using the test datasource
<details>
<summary>Dashboard JSON</summary>
{
  "annotations": {
    "list": [
      {
        "builtIn": 1,
        "datasource": {
          "type": "datasource",
          "uid": "grafana"
        },
        "enable": true,
        "hide": true,
        "iconColor": "rgba(0, 211, 255, 1)",
        "name": "Annotations & Alerts",
        "type": "dashboard"
      }
    ]
  },
  "editable": true,
  "fiscalYearStartMonth": 0,
  "graphTooltip": 0,
  "id": 1,
  "links": [],
  "panels": [
    {
      "datasource": {
        "type": "grafana-testdata-datasource",
        "uid": "ceaqw2sfltr7kb"
      },
      "description": "",
      "fieldConfig": {
        "defaults": {
          "color": {
            "mode": "palette-classic"
          },
          "custom": {
            "axisBorderShow": false,
            "axisCenteredZero": false,
            "axisColorMode": "text",
            "axisLabel": "",
            "axisPlacement": "auto",
            "barAlignment": 0,
            "barWidthFactor": 0.6,
            "drawStyle": "points",
            "fillOpacity": 0,
            "gradientMode": "none",
            "hideFrom": {
              "legend": false,
              "tooltip": false,
              "viz": false
            },
            "insertNulls": false,
            "lineInterpolation": "linear",
            "lineWidth": 1,
            "pointSize": 5,
            "scaleDistribution": {
              "type": "linear"
            },
            "showPoints": "auto",
            "spanNulls": false,
            "stacking": {
              "group": "A",
              "mode": "none"
            },
            "thresholdsStyle": {
              "mode": "off"
            }
          },
          "mappings": [],
          "thresholds": {
            "mode": "absolute",
            "steps": [
              {
                "color": "green",
                "value": null
              },
              {
                "color": "red",
                "value": 80
              }
            ]
          }
        },
        "overrides": []
      },
      "gridPos": {
        "h": 6,
        "w": 24,
        "x": 0,
        "y": 0
      },
      "id": 2,
      "options": {
        "graph": {
          "showBars": false,
          "showLines": true,
          "showPoints": false
        },
        "legend": {
          "calcs": [],
          "displayMode": "list",
          "placement": "bottom",
          "showLegend": true
        },
        "series": {},
        "tooltip": {
          "mode": "single",
          "sort": "none"
        }
      },
      "pluginVersion": "11.4.0",
      "targets": [
        {
          "datasource": {
            "type": "grafana-testdata-datasource",
            "uid": "ceaqw2sfltr7kb"
          },
          "refId": "A",
          "scenarioId": "streaming_client",
          "stream": {
            "noise": 10,
            "speed": 100,
            "spread": 20,
            "type": "signal"
          },
          "stringInput": ""
        }
      ],
      "title": "Simple dummy streaming example",
      "type": "timeseries"
    }
  ],
  "preload": false,
  "refresh": "",
  "schemaVersion": 40,
  "tags": [],
  "templating": {
    "list": []
  },
  "time": {
    "from": "now-1m",
    "to": "now"
  },
  "timepicker": {},
  "timezone": "",
  "title": "Simple Streaming Example",
  "uid": "TXSTREZ",
  "version": 2,
  "weekStart": ""
}
</details>

Executing the test in this szenario works as expected 
```
=== RUN   TestDifferentInstanceURL
--- PASS: TestDifferentInstanceURL (0.01s)
PASS
ok      github.com/mperkins808/grafanadata/go/pkg/grafanadata   0.018s
```

Changing the setup:
1.  Run grafana on port 3000 using `docker run -d --name=grafana3000 -p 3000:3000  -e "GF_SERVER_ROOT_URL=http://localhost/grafana"` grafana/grafana
2.  Create reverse proxy from localhost/grafana to localhost:3000 using apache2 (requries `a2enmod proxy_http`)
```
<VirtualHost *:80>
	ErrorLog ${APACHE_LOG_DIR}/error.log
	CustomLog ${APACHE_LOG_DIR}/access.log combined
	ProxyPreserveHost On
	ProxyPass /grafana http://localhost:3000
	ProxyPassReverse /grafana http://localhost:3000
</VirtualHost>
```
3. Create DataSource, Service account with Access Token and Dashboards like in the previous test
4. Running the test fails
```
=== RUN   TestDifferentInstanceURL
    /home/huber/coding/github/grafanadata/go/pkg/grafanadata/grafanadata_test.go:173: grafana returned status 404
--- FAIL: TestDifferentInstanceURL (0.00s)
FAIL
FAIL    github.com/mperkins808/grafanadata/go/pkg/grafanadata   0.005s
```

The request is made to `http://localhost/api/dashboards/uid/TXSTREZ` instead of `http://localhost/grafana/api/dashboards/uid/TXSTREZ`
Manually testing `curl -H "Authorization: Bearer glsa_[my_token]" http://localhost/grafana/api/dashboards/uid/TXSTREZ` sucessfully returns the dashboard data

--- a/README.md
+++ b/README.md
@@ -22,8 +22,12 @@ import (
 
 func main() {
 
-	u := "http://localhost:3000/d/bebca380-068d-463d-9c9c-1bb19cb8d2b3/new-dashboard?orgId=1&viewPanel=2"
+	u := "http://localhost:3000/"
 	t := "glsa_5N21WQvXza0oWkbqQvjOhII8yJYxGS0G_fbb82943"
+	uid := "bebca380-068d-463d-9c9c-1bb19cb8d2b3"
+	panelID := 7
+
+
 	client, err := grafanadata.NewGrafanaClient(u, t)
 
 	if err != nil {
@@ -31,7 +35,7 @@ func main() {
 	}
 
 	start := time.Now().Add(time.Hour * 24 * -2)
-	data, err := client.GetPanelDataFromURL(u, start)
+	data, err := client.GetPanelDataFromID(uid, panelID, start)
 
 	if err != nil {
 		log.Fatal(err)
@@ -71,7 +75,7 @@ func main() {
 
 	uid := "bebca380-068d-463d-9c9c-1bb19cb8d2b3"
 
-	resp, err := client.GetDashboardWithUID(uid)
+	resp, err := client.GetDashboard(uid)
 	if err != nil {
 		return
 	}

--- a/go/pkg/grafanadata/dashboard.go
+++ b/go/pkg/grafanadata/dashboard.go
@@ -5,11 +5,14 @@ import (
 	"fmt"
 	"io"
 	"net/http"
+	"strings"
 )
 
 // returns all the dashboards for a grafana instance
 func (c *grafanaClient) FetchDashboards() ([]DashboardSearch, error) {
-	q := fmt.Sprintf("%v://%v/api/search?type=dash-db", c.baseURL.Scheme, c.baseURL.Host)
+	host := strings.TrimSuffix(c.baseURL.String(), "/")
+
+	q := fmt.Sprintf("%v/api/search?type=dash-db", host)
 	req, err := c.NewRequest(http.MethodGet, q, nil)
 	if err != nil {
 		return nil, err

--- a/go/pkg/grafanadata/grafanadata.go
+++ b/go/pkg/grafanadata/grafanadata.go
@@ -17,12 +17,11 @@ type GrafanaClient interface {
 	SetToken(token string)
 	NewRequest(method, endpoint string, body io.Reader) (*http.Request, error)
 	Do(req *http.Request) (*http.Response, error)
-	GetDashboardFromURL(urlstr string) (GrafanaDashboardResponse, error)
-	GetDashboardWithUID(uid string) (GrafanaDashboardResponse, error)
-	GetPanelDataFromURL(urlstr string, start time.Time) (Results, error)
+	GetDashboard(uid string) (GrafanaDashboardResponse, error)
 	GetPanelDataFromID(uid string, panelID int, start time.Time) (Results, error)
 	FetchDashboards() ([]DashboardSearch, error)
 	FetchPanelsFromDashboard(dashboard GrafanaDashboardResponse) []PanelSearch
+	GetHost() string
 }
 
 // needed for unit tests
@@ -72,7 +71,7 @@ func (c *grafanaClient) SetToken(token string) {
 func (c *grafanaClient) getDashboard(uid string) (GrafanaDashboardResponse, error) {
 	var grafanaDashboardResponse GrafanaDashboardResponse
 
-	host := fmt.Sprintf("%v://%v", c.baseURL.Scheme, c.baseURL.Host)
+	host := strings.TrimSuffix(c.baseURL.String(), "/")
 	query := fmt.Sprintf("%v/api/dashboards/uid/%v", host, uid)
 
 	req, err := c.NewRequest(http.MethodGet, query, nil)
@@ -129,7 +128,8 @@ func (c *grafanaClient) getPanelData(panelID int, dashboard GrafanaDashboardResp
 		return result, fmt.Errorf("failed to build request object %v", err)
 	}
 
-	query := fmt.Sprintf("%v://%v/api/ds/query", c.baseURL.Scheme, c.baseURL.Host)
+	host := strings.TrimSuffix(c.baseURL.String(), "/")
+	query := fmt.Sprintf("%v/api/ds/query", host)
 	req, err := c.NewRequest(http.MethodPost, query, bytes.NewBuffer(b))
 	if err != nil {
 		return result, fmt.Errorf("failed to build request %v", err)
@@ -154,36 +154,9 @@ func (c *grafanaClient) getPanelData(panelID int, dashboard GrafanaDashboardResp
 	return result, err
 }
 
-// retrieves a dashboard object from the URL
-func (c *grafanaClient) GetDashboardFromURL(urlstr string) (GrafanaDashboardResponse, error) {
-	var grafanaDashboardResponse GrafanaDashboardResponse
-	uid, _ := ExtractArgs(urlstr)
-
-	if uid == "" {
-		return grafanaDashboardResponse, fmt.Errorf("failed to extract uid from url")
-	}
-
-	grafanaDashboardResponse, err := c.getDashboard(uid)
-	return grafanaDashboardResponse, err
-}
-
 // retrieves a dashboard object from a uid
-func (c *grafanaClient) GetDashboardWithUID(uid string) (GrafanaDashboardResponse, error) {
+func (c *grafanaClient) GetDashboard(uid string) (GrafanaDashboardResponse, error) {
 	return c.getDashboard(uid)
-}
-
-// retrieves the panel data from a URL
-func (c *grafanaClient) GetPanelDataFromURL(urlstr string, start time.Time) (Results, error) {
-	uid, id := ExtractArgs(urlstr)
-
-	var result Results
-	dashboard, err := c.getDashboard(uid)
-	if err != nil {
-		return result, err
-	}
-
-	result, err = c.getPanelData(id, dashboard, start)
-	return result, err
 }
 
 // retrieves the panel data from an id
@@ -245,4 +218,8 @@ func ExtractArgs(urlStr string) (string, int) {
 	}
 
 	return uid, int(id)
+}
+
+func (c *grafanaClient) GetHost() string {
+	return strings.TrimSuffix(c.baseURL.String(), "/")
 }

--- a/go/pkg/grafanadata/grafanadata_test.go
+++ b/go/pkg/grafanadata/grafanadata_test.go
@@ -6,6 +6,7 @@ import (
 	"net/http"
 	"net/url"
 	"os"
+	"strings"
 	"testing"
 	"time"
 )
@@ -48,15 +49,11 @@ func CreateMockGrafanaClient(t *testing.T, mockClient *MockHTTPClient) *grafanaC
 
 func TestCreateGrafanaClient(t *testing.T) {
 
-	c, err := NewGrafanaClient("foo", "bar")
+	_, err := NewGrafanaClient("foo", "bar")
 	if err != nil {
 		t.Fatalf("creating new grafana client error %v", err)
 	}
 
-	_, err = c.GetDashboardFromURL("foo")
-	if err == nil {
-		t.Fatalf("should have failed with bad url")
-	}
 }
 
 func TestGetDashboard(t *testing.T) {
@@ -156,5 +153,29 @@ func TestGetDashboards(t *testing.T) {
 	_, err = g.getDashboard("foo")
 	if err == nil {
 		t.Fatal("wanted error but was nil")
+	}
+}
+
+func TestVariousGrafanaUrls(t *testing.T) {
+	token := "foo"
+	urls := []string{
+		"http://grafana:3000",
+		"http://grafana:3000/",
+		"http://grafana/",
+		"http://grafana",
+		"https://host/grafana/",
+		"https://host/grafana",
+	}
+
+	for _, url := range urls {
+		client, err := NewGrafanaClient(url, token)
+		if err != nil {
+			t.Fatalf("could not created GrafanaClient from %v: %v", url, err)
+		}
+
+		host := client.GetHost()
+		if host != strings.TrimSuffix(url, "/") {
+			t.Errorf("expected %v. got %v", url, host)
+		}
 	}
 }

